src/v2/providers/pubsub.ts

```diff
@@ -0,0 +1,158 @@
+  /**
+   * Autogenerated ID that uniquely identifies this message.
+   */
+  readonly messageId: string;
+
+  /**
+   * Time the message was published
+   */
+  readonly publishTime: string;
+
+  /**
+   * The data payload of this message object as a base64-encoded string.
+   */
+  readonly data: string;
+
+  /**
+   * User-defined attributes published with the message, if any.
+   */
+  readonly attributes: { [key: string]: string };
+
+  /**
+   * User-defined key used to ensure ordering amongst messages with the same key.
+   */
+  readonly orderingKey: string;
+
+  /** @hidden */
+  private _json: any;
+
+  constructor(data: any) {
+    this.messageId = data.messageId;
+    this.data = data.data;
+    (this.attributes = data.attributes || {}), (this._json = data.json);
+    this.orderingKey = data.orderingKey;
+    this.publishTime = data.publishTime || new Date().toISOString();
+  }
+
+  /**
+   * The JSON data payload of this message object, if any.
+   */
+  get json(): any {
+    if (typeof this._json === 'undefined') {
+      this._json = JSON.parse(
+        Buffer.from(this.data, 'base64').toString('utf8')
+      );
+    }
+
+    return this._json;
+  }
+
+  /**
+   * Returns a JSON-serializable representation of this object.
+   *
+   * @return A JSON-serializable representation of this object.
+   */
+  toJSON(): any {
+    const json: Record<string, any> = {
+      messageId: this.messageId,
+      data: this.data,
+      publishTime: this.publishTime,
+    };
+    if (Object.keys(this.attributes).length) {
+      json.attributes = this.attributes;
+    }
+    if (this.orderingKey) {
+      json.orderingKey = this.orderingKey;
+    }
+    return json;
+  }
+}
+
+/** The interface published in a Pub/Sub publish subscription. */
+export interface MessagePublishedData {
+  readonly message: Message;
+  readonly subscription: string;
+}
+
+/** PubSubOptions extend EventHandlerOptions but must include a topic. */
+export interface PubSubOptions extends options.EventHandlerOptions {
+  topic: string;
+}
+
+/** Handle a message being published to a Pub/Sub topic. */
+export function onMessagePublished(
+  topic: string,
+  callback: (event: CloudEvent<MessagePublishedData>) => any | Promise<any>
+): CloudFunction<MessagePublishedData>;
+
+/** Handle a message being published to a Pub/Sub topic. */
+export function onMessagePublished(
+  options: PubSubOptions,
+  callback: (event: CloudEvent<MessagePublishedData>) => any | Promise<any>
+): CloudFunction<MessagePublishedData>;
+
+export function onMessagePublished(
+  topicOrOptions: string | PubSubOptions,
+  callback: (event: CloudEvent<MessagePublishedData>) => any | Promise<any>
+): CloudFunction<MessagePublishedData> {
+  let topic: string;
+  let opts: options.EventHandlerOptions;
+  if (typeof topicOrOptions === 'string') {
+    topic = topicOrOptions;
+    opts = {};
+  } else {
+    topic = topicOrOptions.topic;
+    opts = { ...topicOrOptions };
+    delete (opts as any).topic;
```

Comment:
I think we want to establish scalable patterns, and I'd rather be fully confident that unwanted fields are discarded by `optionsToTriggerAnnotations` than start a pattern of unnecessarily removing these fields. It's simple in this case, but when we do audit logs there may be 3-4 different event-specific options and it could become a bit much.

Not going to block approval on it, but that's my 2c